### PR TITLE
fix: sliding window KV cache for Gemma-3 models (issue #2145)

### LINKED ISSUE #2145: litgpt model responses using simple "out-of-box" code example become incoherent / repetitive after a few hundred tokens
### Bug description

I’m using litgpt (Version: 0.5.11) to chat with the following models:

checkpoints/google/gemma-3-4b-it
checkpoints/meta-llama/Llama-3.2-3B-Instruct

I deliberately prompt both models to generate a long response and I set the max_new_tokens generation parameter to 1500 tokens. In response to this single prompt, both models start out giving a coherent response but after a few hundred generated words, both models begin generating either gibberish or they output repeated instances of the same sentence/word/phrase.  I've been able to consistently reproduce this behavior with different random seeds.

I realize this may not be a bug in litgpt itself (perhaps a misconfiguration), but I see this problem when I try to use litgpt "out-of-box" using code (see attached file runlitgpt.py) that is close to the minimal example described [in this litgpt API documentation](https://github.com/Lightning-AI/litgpt/blob/main/tutorials/python-api.md#generatechat).  Also, I am using values for temperature, top_p, and top_k that are recommended for each model (gemma3 and llama-3.2).  Even if there's a misconfiguration, it seems odd that the models would be so misconfigured out of the box.

Clearly, this behavior isn’t expected since these models have been trained for long contexts and I haven’t seen this behavior when running the same models in frameworks other than Lightning/litgpt.  Examples where I've successfully run the same prompts using the same generation parameters (temperature, top_p, top_k) include [Google's pytorch inference implementation of Gemma3](https://github.com/google/gemma_pytorch/blob/main/scripts/run_multimodal.py) and the HuggingFace transformers implementation of Llama3.2.

I'm including some text file dumps of both prompts and model responses to illustrate the problem.  The file names encode the model used, the random seed, and the generation parameter arguments supplied in calls to llm.generate():  max_new_tokens, temperature, top_p, top_k.

[gemma_seed_0_promptindex_0_maxtokens_1500_temp_1.0_topk_64_topp_0.95.txt](https://github.com/user-attachments/files/22991230/gemma_seed_0_promptindex_0_maxtokens_1500_temp_1.0_topk_64_topp_0.95.txt)
[gemma_seed_0_promptindex_1_maxtokens_1500_temp_1.0_topk_64_topp_0.95.txt](https://github.com/user-attachments/files/22991231/gemma_seed_0_promptindex_1_maxtokens_1500_temp_1.0_topk_64_topp_0.95.txt)
[gemma_seed_0_promptindex_2_maxtokens_1500_temp_1.0_topk_64_topp_0.95.txt](https://github.com/user-attachments/files/22991228/gemma_seed_0_promptindex_2_maxtokens_1500_temp_1.0_topk_64_topp_0.95.txt)
[llama_seed_0_promptindex_0_maxtokens_1500_temp_0.6_topk_64_topp_0.9.txt](https://github.com/user-attachments/files/22991232/llama_seed_0_promptindex_0_maxtokens_1500_temp_0.6_topk_64_topp_0.9.txt)
[llama_seed_0_promptindex_1_maxtokens_1500_temp_0.6_topk_64_topp_0.9.txt](https://github.com/user-attachments/files/22991233/llama_seed_0_promptindex_1_maxtokens_1500_temp_0.6_topk_64_topp_0.9.txt)
[llama_seed_0_promptindex_2_maxtokens_1500_temp_0.6_topk_64_topp_0.9.txt](https://github.com/user-attachments/files/22991229/llama_seed_0_promptindex_2_maxtokens_1500_temp_0.6_topk_64_topp_0.9.txt)

Here is a simple python script that can be used to reproduce the problem:

[runlitgpt.py](https://github.com/user-attachments/files/22991267/runlitgpt.py)

### Reproduced in studio

_No response_

### What operating system are you using?

Linux

### LitGPT Version

Output of pip show litgpt | grep Version:   Version: 0.5.11

Python file I am running on my Linux system which can be used to reproduce the problem:

[runlitgpt.py](https://github.com/user-attachments/files/22991251/runlitgpt.py)

### PR DESCRIPTION
**Pull request by Chelsi-create**
fixes #2145 
### Overview
This PR addresses a critical issue affecting **Gemma-3 models** (`4B-IT`, `12B-IT`, `27B-IT`) that caused them to produce gibberish or repetitive text after approximately **800-1000 tokens** during continuous long-form generation.  
The fix introduces **correct sliding window KV cache management** for models using hybrid attention architectures.

#### 1. Sliding Window KV Cache Limiting (`litgpt/model.py` - `build_kv_cache`)
- Use `sliding_window_size` (1024 tokens) for sliding-window layers instead of full sequence length.
- Allocate correct cache size during initialization.
- Global attention layers continue to use the full-sequence cache.

#### 2. Circular Buffer Implementation (`litgpt/model.py` - `KVCache` class)
- Implement modulo-based position indexing for sequences beyond the window size.
- Ensure the KV cache never exceeds 1024 tokens for sliding-window layers.

#### 3. Attention Mask Dimension Fix (`litgpt/model.py` - `CausalSelfAttention.forward`)
- Adjust attention mask dimensions to align with actual KV cache sizes.
- Prevents dimension mismatch errors when cache is smaller than the expected sequence length.

#### The Hidden Bug
This issue was difficult to detect in production:
- **Interactive chat sessions (100–300 tokens)** always reset the KV cache between turns, so no visible problem.
- **Continuous generation (>1024 tokens)** triggered the bug, leading to repetitive text.

**Testing Summary:**
- `Gemma-3-4B-IT`: All three long-form test prompts passed.
- `Llama-3.2-3B`: All three long-form test prompts passed.
- Before fix: severe repetition (150-200 token loops).  
- After fix: coherent outputs with limited repetition (7-9 tokens).

**Configuration:**
- API: `chat_generate()` (low-level)
- `max_new_tokens=1500`

### Known Limitations
- `LLM.generate()` (high-level API) still requires integration work.
- Low-level `chat_generate()` API is fully functional.
- Temporary workaround: use `chat_generate()` directly (example in full PR description).

### Notes
- **Single file modified:** `litgpt/model.py` (~40 lines)
- **No breaking changes:** Fully backward compatible
- **Validated:** Affected models tested

[runlitgpt_v1.py](https://github.com/user-attachments/files/23287612/runlitgpt_v1.py) - this file uses the high-level API implementation. There are atill some bugs in the `litgpt/api.py`, which will be addresses in another PR.
[runlitgpt.py](https://github.com/user-attachments/files/23287613/runlitgpt.py) - this code uses the low-level API implementation and it works for both the models now.

### CODE REVIEW

**Comment by KaelanDt** [comment]
Hi @Chelsi-create , thank you for the PR!
There are still a few tests failing related to your changes, could you check that the tests pass?
https://github.com/Lightning-AI/litgpt/actions/runs/19088536192/job/54534011522?pr=2153

**Comment by drwslacy47** [comment]
Hi @Chelsi-create, thanks very much for working on this!  I am the original filer of the bug and I really appreciate you taking up the issue.  I have several questions related to your testing and findings.

**(1)** It looks like you are focused on an issue with sliding window attention in Gemma, but in my original runlitgpt.py script (using LLM.generate), I also see the same problem of repetition when running Llama.  **Were you not able to replicate the problem with Llama using my  runlitgpt.py script?**  And if you were able to reproduce the problem with Llama, did this problem get somehow fixed as a side effect of the fix you made for sliding window KV cache management in Gemma?  I realize that Llama only uses Global Attention, so that's why I am confused how the problem with Llama was also apparently fixed (you mention Llama passing all three "long form" generation tests in your testing summary).

Another thing worth mentioning here ... The problem with Gemma (before your fix) occurs both when using LLM.generate (as in my runlitgpt.py script) and also when running Gemma using **litgpt chat** (for the same prompts and parameter settings).  However, when running Llama, I only see the repetition problem when running Llama with LLM.generate and **NOT** when I use the same prompt and parameters in litgpt chat.  **This suggests there are two different bugs.**  Gemma probably fails in both cases because of the sliding window context issue you are addressing with this PR.  **However, why would Llama fail when using LLM.generate (with very similar symptoms as the Gemma failure) but not fail when using litgpt chat?**  This is all quite puzzling to me. Of course, my comments here assume you are able to reproduce what I just described:  Llama has the repetitive behavior (just like Gemma) before your fix when using LLM.generate() but **not** when using litgpt chat.

**(2)** You make a distinction between LLM.generate and chat_generate(), but your linked files for runlitgpt.py and runlitgpt_v1.py both seem to be using LLM.generate. The only difference seems to be in the prompts.  So, I am a little confused about the difference between LLM.generate and chat_generate() and why this distinction matters in the context of this bug.

**(3)**  You mention this in your testing summary:  "After fix: coherent outputs with limited repetition (7-9 tokens)."  I'm  somewhat concerned to hear that there is still any repetition at all.  Can you share an example of the "limited repetition" you are seeing? 

I'd prefer we not close the original issue until these questions are addressed.  Please let me know if there is any way I can help in resolving any of these questions.  I'd really like to use litgpt but I'm uneasy about that until I understand the answers to the questions I raised above.

Thanks!

**Comment by Chelsi-create** [comment]
Hi @drwslacy47 ,

Thank you so much for the detailed and thoughtful questions. I have tried to address all your questions below:

### (1) Llama Behavior: Why it Failed with `LLM.generate()` but Works with `chat_generate()`
This is an excellent observation, and I should clarify what I discovered.

**The Core Finding**
After careful investigation, both Gemma and Llama fail with the high-level LLM.generate() API (as in your original runlitgpt_v1.py), but the low-level chat_generate() based runlitgpt.py works for both models.

Your original runlitgpt_v1.py using llm.generate() → Shows repetition for Llama
The runlitgpt.py using chat_generate() → Works coherently for Llama

**Why does this happen?**
The difference is two separate issues:

For Gemma:
Primary issue: Sliding window KV cache bug (what this PR fixes)
**When using chat_generate() + correct stop tokens → Works perfectly**
When using LLM.generate() → Still fails due to cache management in api.py

For Llama:
No sliding window issue (uses global attention only)
Problem: LLM.generate() in api.py line 762 uses only EOS token for stopping
**When using chat_generate() + proper model stop tokens (prompt_style.stop_tokens()) -> this works**
When using LLM.generate() + only EOS token → Model keeps generating past natural stopping point

**Why Works in litgpt chat but **not** in LLM.generate()?**

- The litgpt chat CLI uses the same chat_generate() function as `runlitgpt.py` script:
- It correctly uses `prompt_style.stop_tokens()`
- Each user turn resets the KV cache
- No long-form generation in single call

**Core problem with the LLM.generate() API:**

- Uses incomplete stop token handling (only EOS) 
- Tries to manage cache internally without proper reset 
- Has different code path in api.py that bypasses my model.py fix partially


**The Key Difference Between the Two Scripts**
runlitgpt_v1.py (BROKEN - uses LLM.generate()):

python
```
response = llm.generate(
    prompt,
    max_new_tokens=1500,
    temperature=0.6,
    top_k=64,
    top_p=0.9
)
```
runlitgpt.py (WORKING - uses chat_generate()):

python
```
from litgpt.chat.base import generate as chat_generate

stop_tokens = prompt_style.stop_tokens(tokenizer)  # ← Gets ALL stop tokens!
stream = chat_generate(
    model=llm.model,
    prompt=input_ids,
    max_returned_tokens=max_returned_tokens,
    temperature=0.6,
    top_k=64,
    top_p=0.9,
    stop_tokens=stop_tokens  # ← Uses proper stop tokens
)
```

**Short summary:**
The runlitgpt.py script works for BOTH models because:

- Properly pre-sizes KV cache with llm.model.set_kv_cache()
- Uses model-specific stop tokens: prompt_style.stop_tokens(tokenizer)
- Uses low-level chat_generate() directly
- The sliding window fix in model.py is properly utilized

### (2) Clarifying LLM.generate() vs chat_generate()

**LLM.generate() - High-Level Python API**
Location: litgpt/generate/api.py
Advantages: Simple, one-line API
Disadvantages: Limited stop token control, internal cache management issues

**chat_generate() - Low-Level Function**
Advantages: Complete control, explicit stop tokens, cleaner cache handling
Disadvantages: More setup required

Why This Matters for the Bug
Your original script uses LLM.generate():

python
```
llm = LLM.load(model_path)
response = llm.generate(prompt, ...)  # High-level API
```

python
```
from litgpt.chat.base import generate as chat_generate

stop_tokens = prompt_style.stop_tokens(tokenizer)  # ← Explicitly get all stop tokens
stream = chat_generate(..., stop_tokens=stop_tokens)
```
So both your runlitgpt_v1.py and runlitgpt.py conceptually do generation, but runlitgpt.py actually uses the lower-level chat_generate() underneath, which gives us full control.

### (3) The "Limited Repetition (7-9 tokens)" Explanation

My earlier statement about “limited repetition” was imprecise. What I actually measured was word frequency in the last 200 tokens, which isn’t a good indicator of coherence. Natural repetition (eg “the”, “and”) can appear frequently without being an issue.

**Comment by drwslacy47** [comment]
Hi @Chelsi-create.  Thanks for your answers to my questions -- I really appreciate you taking time to respond with those details.  I think I understand what you have related. I will make another pass over your comments to be sure I appreciate the difference in the two issues. I need to understand the use of prompt_style.stop_tokens() better.  

Another quick question: Is there an ETA on when your fix would be available? It looks like your pull request still needs an approval?

Again, thank you for working on this issue and for responding to my questions. I apologize for the delay in responding to your latest comment.

-- Steve

**Comment by lianakoleva** [comment]
@Chelsi-create Thank you so much for the contribution! Can you take a look at the failing tests? Let me know if you need some help

**Comment by Chelsi-create** [comment]
Hi @lianakoleva 
Thank you so much! I took a look, but I’ve currently lost access to GPUs on the HPC cluster. I should be getting access back during winter break, and once that’s sorted, I’ll be able to properly debug and fix the failing tests.
I’ll try to get to this as soon as possible after that. Please let me know if that timeline works on your end.
Thanks for your understanding.

**Comment by lianakoleva** [comment]
hey no worries, i already pushed the needed fix @Chelsi-create head over to https://lightning.ai for free credits for GPUs refreshed every month (:
